src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.cryostat.discovery;
+
+import java.net.URI;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.management.remote.JMXServiceURL;
+
+import io.cryostat.core.sys.FileSystem;
+import io.cryostat.targets.Target;
+import io.cryostat.targets.Target.Annotations;
+
+import io.fabric8.kubernetes.api.model.EndpointAddress;
+import io.fabric8.kubernetes.api.model.EndpointPort;
+import io.fabric8.kubernetes.api.model.EndpointSubset;
+import io.fabric8.kubernetes.api.model.Endpoints;
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.ObjectReference;
+import io.fabric8.kubernetes.api.model.OwnerReference;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import io.quarkus.runtime.ShutdownEvent;
+import io.quarkus.runtime.StartupEvent;
+import io.smallrye.mutiny.infrastructure.Infrastructure;
+import jakarta.annotation.Priority;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.event.Observes;
+import jakarta.inject.Inject;
+import jakarta.persistence.NoResultException;
+import jakarta.transaction.Transactional;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.concurrent.ConcurrentException;
+import org.apache.commons.lang3.concurrent.LazyInitializer;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.commons.lang3.tuple.Triple;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.logging.Logger;
+
+@ApplicationScoped
+public class KubeApiDiscovery {
+    public static final String REALM = "KubernetesApi";
+
+    public static final long ENDPOINTS_INFORMER_RESYNC_PERIOD = Duration.ofSeconds(30).toMillis();
+
+    public static final String DISCOVERY_NAMESPACE_LABEL_KEY = "discovery.cryostat.io/namespace";
+
+    @Inject Logger logger;
+
+    @Inject KubeConfig kubeConfig;
+
+    @ConfigProperty(name = "cryostat.discovery.kubernetes.enabled")
+    boolean enabled;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-names")
+    Optional<List<String>> JmxPortNames;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-numbers")
+    Optional<List<Integer>> JmxPortNumbers;
```

Comment:
What happens if you just declare this as `List<Integer>` and have `cryostat.discovery.kubernetes.port-numbers=` in `application.properties`? IIRC the config framework will give you an empty list already in this case. I think the current behaviour you've implemented is already correct, so I just wonder if the `Optional<>` can be cleaned up a little this way.

Reply:
I guess `Optional` is necessary then. If I remove it, there is an error at startup:

```
Failed to load config value of type java.util.List<java.lang.String> for: cryostat.discovery.kubernetes.namespaces
```

Reply:
Yeah it seems so...it works fine with non-empty values tho. Just when it's empty, there is an error above :((